### ISSUE TITLE
two y axis with different coordinates, one on left and one on right

### ISSUE THREAD

**Issue by asmwarrior** [GitHubLionel/wxMathPlot#20]
Hi, I have a feature request which I would like to ask here.

In my application, I need to plot two plots in a same graph/chart. But each plot has its own unit.

This means that I need two Y axis, one on left margin and one on right margin.

Is this possible? I'm not sure you can understand my need, maybe, this is something like in the below link:

[Plots with different scales — Matplotlib 3.8.3 documentation](https://matplotlib.org/stable/gallery/subplots_axes_and_figures/two_scales.html)

Thanks.

**Comment by GitHubLionel** [comment]
Hi,

I think this is not possible. You can create a second Y axis but you can not attach it to a specific plot.
And I don't know if it is possible to do that without major change in the code.
It would be necessary to be able to connect a graph to an axis. 
Interesting problem but no time to investigate that.
Regards,
Lionel

**Comment by asmwarrior** [comment]
Hi, Lionel, thanks for your response.

I know it is complex to add such feature, so I think leave this issue open. Maybe some day we can find a way to implement it.

Thanks.

**Comment by asmwarrior** [comment]
Some good news about two Y axies in Wisteria-Dataviz (also a wxWidgets based project), see here:

[Any feature about two Y axies in the chart? · Issue #43 · Blake-Madden/Wisteria-Dataviz](https://github.com/Blake-Madden/Wisteria-Dataviz/issues/43)

**Comment by asmwarrior** [comment]
I have one question, for the code here:

https://github.com/GitHubLionel/wxMathPlot/blob/93a3405d3fb03e2eb72ae5c8cb8f070a9e587b67/mathplot/mathplot.cpp#L1757-L1859

If I have a second Y(in the right margin) axis, and if I set the maxY value and minY value for this Y axis, and can I calculate each ticks and each labels for this second Y axis?

For example, I have the first Y axis on the window, which is in the Y coordinate range [0, 100], and for the second Y axis, it ranges at [0, 200]. So, for each ticks label in the right axis is twice as the each ticks in the left axis.

I mean, is it possible to customize the tick labels in the right Y axis?

I think the above is the first step. If I can correctly draw the second Y axis, than, I will find a way to draw the lines according to the "scale and shift" of the second Y axis, because I can easily apply the transforms on the lines (x,y) coordinates.

**Comment by GitHubLionel** [comment]
Hi,

For your example, on can simply introduce a new property factor in the mpScaleY class. Then in line 1838 you will have :
`s.Printf(fmt, n * factor);`
So with factor = 2 for the second Y axis, when you read 100 in the first axis, you will read 200 in the second axis.

But well, I don't think that's what you want. 
In fact, I don't know how to set maxY and minY for an axis. The logic is that axis are drawn first according the full screen available (m_plotBondaries). Then after, curves are drawn but there are no link between axis and curves, only the "plotBondaries" is pertinent.

**Comment by asmwarrior** [comment]
I have commit my first changes to my fork ([* try to implement a second Y axis which has different tick labels](https://github.com/asmwarrior/wxMathPlot/commit/3225e618593f7a3fd6a3a84e8c261a5cb064ae68))

here is the result:

![image](https://github.com/GitHubLionel/wxMathPlot/assets/561818/098ae50b-9698-417a-bc97-b233818c83b6)

It looks good.

> In fact, I don't know how to set maxY and minY for an axis. The logic is that axis are drawn first according the full screen available (m_plotBondaries). Then after, curves are drawn but there are no link between axis and curves, only the "plotBondaries" is pertinent.

This is one issue, yes, I see the a curve is not based on an axis, but I think I may need to extend a new curve class, which can depends on the "mpScaleY2".

My design strategy is that I can have a function `newY = oldY * scale + shift` for the Y2 axis, so for the `mpScaleY2` class, I only need to remember the `scale` and `shift` member variable. Now, the remaining steps are:

1, I have already know how to draw the Y axis and Y axis's tick lines(horizontal lines)
2, When initialize the maxY and minY for the Y2 axis, I need to calculate the `scale` and `shift` member variable for the Y2 axis
3, after that, I think the Y2's `scale` and `shift` is totally fixed (depends on the Y axis's maxY and minY)
4, if the left Y axis's max Y and minY get changed, I need to adjust/recalculate the Y2's `scale` and `shift`

The hardest part is the step 2.

Here comes another issue, does the "tick lines" draw twice if I have both Y axis one the left side, and Y2 axis one the right side? How can I determine both the left Y and right Y are drawing in the same "steps" (in pixel coordinates).

**Comment by GitHubLionel** [comment]
Hi,

The crux of the problem is the definition of m_scaleY. All other parameters are defined with it, like m_posY.
So we need another m_scaleY for the second axis.
I made a try. I introduced m_scaleY2, m_posY2 and a factor in mpWindow class. In mpScaleY, I just add a boolean to say that it is another Y axis. And in mpFX, mpFY and mpFXY, another boolean to say that we use another axis.
With these little changes, I have almost solve the problem. 
You just have to :
- define the factor for example to reduce by 2 the axis : m_plot->Y2Factor = 2.0;
- define another axis with Y2Axis to true
- say UseY2Axis to true for the series you want to fit to Y2 axis.

You have sources in the sandbox repository : [https://github.com/GitHubLionel/wxMathPlot/tree/master/Sandbox](url)

There are some problems when you resize the windows or when you select an area, but for the moment, it is just a test.

**Comment by GitHubLionel** [comment]
We have strange behaviour when we select an area to zoom. Since the coordinates are in the first axis referential, if we select an object in the second axis referential, the result is correct but not what we expected. I don't know how to correct this effect.

**Comment by asmwarrior** [comment]
Hi, thanks for your effort, I would suggest that you could use a new "git branch" to handle such feature implementation, because if I try to use the new file inside the "SandBox" folder, it is hard to check the modifications.

I haven't tried your code in the "SandBox" yet, I will test them soon.

About the issue you mentioned:

> We have strange behaviour when we select an area to zoom. Since the coordinates are in the first axis referential, if we select an object in the second axis referential, the result is correct but not what we expected. I don't know how to correct this effect.

I think it is OK to keep a scale and shift value from the coordinates transform based on Y1 axis to Y2 axis. That scale and shift value should be fixed. So, once you zoom the Y1 axis, you should also zoom the Y2 axis in the same level. I mentioned this way in my previous comments.

Thanks.

**Comment by GitHubLionel** [comment]
Yes, if you zoom on Y1, you also zoom in Y2. It is logic.
But if we have this configuration with blue/yellow sinus on Y2 (range 1-2) and others curves on Y1 :
![Z1](https://github.com/GitHubLionel/wxMathPlot/assets/73015867/51a1c6c8-8242-490d-a924-e247ff19d021)
And then try to select an area around it, you will obtain this :
![Z2](https://github.com/GitHubLionel/wxMathPlot/assets/73015867/dfa8f134-c534-44d9-b6f2-3281a73cec2c)
because in fact, you have selected an area in range 2-4 (Y1 axis).
Result is correct (we see green curve) but not what you expected (blue/yellow sinus is not visible).

**Comment by asmwarrior** [comment]
Sorry for the late reply.

I just copy the wxMathPlot.h/cpp file from the SandBox folder to the "mathplot", and can you show me what is the client code to use the second Y axis?

Thanks.

**Comment by GitHubLionel** [comment]
Has I write previously, 
You just have to :
- define the factor for example to reduce by 2 the axis : m_plot->Y2Factor = 2.0;
- define another axis with Y2Axis to true
- say UseY2Axis to true for the series you want to fit to Y2 axis.
So :
m_plot->Y2Factor = 2.0;
mpScaleY *rightAxis = new mpScaleY(wxT("Y2"), mpALIGN_RIGHT, true, true);
For a mpFX, mpFy, ... serie : serie->UseY2Axis = true;

**Comment by asmwarrior** [comment]
Thanks. I see you mentioned issue.

This issue happens you use the mouse to drag a rectangle, and you want to zoom to this rectangle.
This issue does not happen when you use mouse scroll to zoom in or zoom out the scene.

So, my guess is that the issue happens from the function:

~~~~
void mpWindow::ZoomRect(wxPoint p0, wxPoint p1)
{
  // Compute the 2 corners in graph coordinates:
  double p0x = p2x(p0.x);
  double p0y = p2y(p0.y);
  double p1x = p2x(p1.x);
  double p1y = p2y(p1.y);

  // Order them:
  mpFloatRect zoom;
  zoom.Xmin = p0x < p1x ? p0x : p1x;
  zoom.Xmax = p0x > p1x ? p0x : p1x;
  zoom.Ymin = p0y < p1y ? p0y : p1y;
  zoom.Ymax = p0y > p1y ? p0y : p1y;

#ifdef MATHPLOT_DO_LOGGING
  wxLogMessage(_T("Zoom: (%f,%f)-(%f,%f)"), zoom.Xmin, zoom.Ymin, zoom.Xmax, zoom.Ymax);
#endif

  Fit(zoom);
}
~~~~

Here, the `zoom` variable is defined for the view XY coordinates of Y1.

So, later, inside the function `Fit(zoom)`, you need to special handling of the zoom. So that the zoom can correctly apply to the lines depends on Y2 axis.

~~~~
// JL
void mpWindow::Fit(const mpFloatRect &rect, wxCoord *printSizeX, wxCoord *printSizeY)
{
  if (m_magnetize)
  {
    // Avoid paint cross if mouse move
    m_repainting = true;
  }

  // Save desired borders:
  m_desired = rect;

  if (printSizeX != NULL && printSizeY != NULL)
  {
    // Printer:
    SetScreen(*printSizeX, *printSizeY);
  }
  else
  {
    // Normal case (screen):
    int h, w;
    GetClientSize(&w, &h);
    SetScreen(w, h);
  }

  double Ax, Ay;

  Ax = rect.Xmax - rect.Xmin;
  Ay = rect.Ymax - rect.Ymin;

  m_scaleX = ISNOTNULL(Ax) ? m_plotWidth / Ax : 1;
  m_scaleY = ISNOTNULL(Ay) ? m_plotHeight / Ay : 1;
  m_scaleY2 = m_scaleY * Y2Factor;

  if (m_lockaspect)
  {
#ifdef MATHPLOT_DO_LOGGING
    wxLogMessage(_T("mpWindow::Fit()(lock) m_scaleX=%f,m_scaleY=%f"), m_scaleX, m_scaleY);
#endif
    // Keep the lowest "scale" to fit the whole range required by that axis (to actually "fit"!):
    double s = m_scaleX < m_scaleY ? m_scaleX : m_scaleY;
    m_scaleX = s;
    m_scaleY = s;
    m_scaleY2 = s;
  }

  // Adjusts corner coordinates: This should be simply:
  //   m_posX = m_minX;
  //   m_posY = m_maxY;
  // But account for centering if we have lock aspect:
  m_posX = (rect.Xmin + rect.Xmax) / 2 - (m_plotWidth / 2 + m_margin.left) / m_scaleX;
  m_posY = (rect.Ymin + rect.Ymax) / 2 + (m_plotHeight / 2 + m_margin.top) / m_scaleY;
  m_posY2 = (rect.Ymin + rect.Ymax) / 2 + (m_plotHeight / 2 + m_margin.top) / m_scaleY2;

#ifdef MATHPLOT_DO_LOGGING
  wxLogMessage(_T("mpWindow::Fit() m_desired.Xmin=%f m_desired.Xmax=%f  m_desired.Ymin=%f m_desired.Ymax=%f"),
      m_desired.Xmin, m_desired.Xmax, m_desired.Ymin, m_desired.Ymax);
  wxLogMessage(_T("mpWindow::Fit() m_scaleX = %f , m_scrX = %d,m_scrY=%d, Ax=%f, Ay=%f, m_posX=%f, m_posY=%f"),
      m_scaleX, m_scrX, m_scrY, Ax, Ay, m_posX, m_posY);
#endif

  // It is VERY IMPORTANT to DO NOT call Refresh if we are drawing to the printer!!
  // Otherwise, the DC dimensions will be those of the window instead of the printer device
  if (printSizeX == NULL || printSizeY == NULL)
    UpdateAll();
}
~~~~

I'm debugging the above function, and try to see if I can solve the issue.

**Comment by asmwarrior** [comment]
I have my test code in this commit:

[add a second Y axis testing code](https://github.com/asmwarrior/wxMathPlot/commit/e6ede159dd4066ef3813fd0c2f1b7658bd7346d4)

I see that sometimes, the relative position/scale of the two curves depends on Y1 and Y2 axes are changing. From my point of view, they should be fixed.

One issue I see is the "tick line" will becomes interleaved, see below images:

![image](https://github.com/GitHubLionel/wxMathPlot/assets/561818/953b13ce-a387-4a4b-947d-4a2170880331)

When I use the mouse scroll to zoom in, it becomes like below:

![image](https://github.com/GitHubLionel/wxMathPlot/assets/561818/2a7b6297-2273-4443-85a1-2c93edde9219)

**Comment by asmwarrior** [comment]
Suppose curve1 is depends on Y1, and Y1 has the property: `YOffset`, `YScale`.

So, to draw curve1 on the screen(pixel coordinates), I have to:   `YPixel = (Y - YOffset) * YScale`

For curve2 on Y2 axis, when draw on the screen, you have to `Y2Pixel = (Y - Y2Offset) * Y2Scale`

But when you zoom the screen, you have to adjust the `YOffset, YScale, Y2Offset, Y2Scale`. But I think their relative position and scale should be fixed.

Currently, I see the relative position and scale are not fixed.

**Comment by asmwarrior** [comment]
OK, I think the below commit can fix such problem, see here:

[* we should key the relative position and scale of the curves from Y1 and Y2 axes](https://github.com/asmwarrior/wxMathPlot/commit/918f46966454cb78d34e131dd5e295b809acce02)

**Comment by asmwarrior** [comment]
This is the method I use, see the below image:

![image](https://github.com/GitHubLionel/wxMathPlot/assets/561818/689c22b8-3371-49e9-8f9b-24f65dc57a84)

I need to calculate the `new_posY2`.

**Comment by asmwarrior** [comment]
When I try to use the right mouse context menu, and click the "lock aspect" item, I see it still has some zoom issue.

**Comment by GitHubLionel** [comment]
Ok, I add your correction and add a test. Now  it is working when we zoom an area, fit or resize.  It is working also with lock aspect but have again a problem when we unlock aspect. But with a fit, thinks return good.

**Comment by asmwarrior** [comment]
I think one issue remains: how to initialized the initial fit.

I mean for example, we have curve0 and curve1 on Y1, and curve2 and curve 3 and Y2. Now, I think the correct way to make a fit is:

Find the bounding box of "curve0 and curve1", and find the bounding box of "curve2 and curve3". Then, we need to find a initial m_posY and m_posY2. Suppose we need to set the Y2Factor value by myself.

If I remember correctly, when start the program, only the "curve0 and curve1" will fit to the canvas window.

**Comment by GitHubLionel** [comment]
Yes need more reflections !
I have posted some modifications to have clean code.
I also add this property to configuration Series page.

**Comment by asmwarrior** [comment]
Thanks. 

> Yes need more reflections !

What does this sentence mean?

I looked at your code, and I see you have a bool variable `FitWithBound`

I'm not sure what does this variable used for, it looks like in the function:

~~~~
void mpWindow::OnSize(wxSizeEvent &WXUNUSED(event))
{
  // Try to fit again with the new window size:
  FitWithBound = true;
  Fit(m_desired);
#ifdef MATHPLOT_DO_LOGGING
  wxLogMessage(_T("mpWindow::OnSize() m_scrX = %d, m_scrY = %d"), m_scrX, m_scrY);
#endif // MATHPLOT_DO_LOGGING
}
~~~~

And

~~~~
void mpWindow::ZoomRect(wxPoint p0, wxPoint p1)
{
  // Compute the 2 corners in graph coordinates:
  double p0x = p2x(p0.x);
  double p0y = p2y(p0.y);
  double p1x = p2x(p1.x);
  double p1y = p2y(p1.y);

  // Order them:
  mpFloatRect zoom;
  zoom.Xmin = p0x < p1x ? p0x : p1x;
  zoom.Xmax = p0x > p1x ? p0x : p1x;
  zoom.Ymin = p0y < p1y ? p0y : p1y;
  zoom.Ymax = p0y > p1y ? p0y : p1y;

#ifdef MATHPLOT_DO_LOGGING
  wxLogMessage(_T("Zoom: (%f,%f)-(%f,%f)"), zoom.Xmin, zoom.Ymin, zoom.Xmax, zoom.Ymax);
#endif

  FitWithBound = true;
  Fit(zoom);
}
~~~~

In the above two cases, the code goes to the first branch. In other cases, it goes to the else branch.

~~~~
  if (FitWithBound)
  {
    m_posY2 += (m_posY - old_posY)/Y2Factor;
    FitWithBound = false;
  }
  else
    m_posY2 = (rect.Ymin + rect.Ymax) / 2 + (m_plotHeight / 2 + m_margin.top) / m_scaleY2;
~~~~

I'm not understand this logic. Thanks.

**Comment by asmwarrior** [comment]
when I start the demo program, I got such result:

![image](https://github.com/GitHubLionel/wxMathPlot/assets/561818/85024725-de09-4a61-acd1-259381c5a987)

It looks like the X coordinates is a combination of two curves(the blue curve0 and the red curve1). While the curve1 is truncated.

The expect result could be like below:
![image](https://github.com/GitHubLionel/wxMathPlot/assets/561818/e5c2f5ec-0250-4b8f-8753-68e631edd782)

Currently, the expect result can be achieved by using the mouse to drag a rectangle which covers the two curves.

Another issue is the tick lines, it is expected that the Y1 and Y2 tick lines should be identical(overlap). I'm not sure how to implement this.

**Comment by GitHubLionel** [comment]
In last post, I try another approach by computing bound for Y and Y2.
Now, Y2Factor is unnecessary.
FitWithBound is a "workaround" for the moment.

**Comment by asmwarrior** [comment]
> In last post, I try another approach by computing bound for Y and Y2. Now, Y2Factor is unnecessary. FitWithBound is a "workaround" for the moment.

You "last post" means the last commit you pushed minutes ago?

[Compute bound on Y and Y2](https://github.com/GitHubLionel/wxMathPlot/commit/1be65f06253e1bca61bf75db33e94e4149d866d2)?

**Comment by GitHubLionel** [comment]
Yes

**Comment by asmwarrior** [comment]
OK, I just tested that commit, it works nice for my example, see image below:

![image](https://github.com/GitHubLionel/wxMathPlot/assets/561818/607989b6-d16b-44a4-82f6-29ed216221b6)

And here is the test code:
~~~~
	mpFXYVector *serie2 = m_plot->GetXYSeries(1);
    for (int i = 0; i < 100; i++)
		serie2->AddData(i / 10.0 + 5,  sin(i / 10.0), true);
    serie2->SetY2Axis(true);
    serie2->SetBrush(*wxRED);
	serie2->SetSymbol(mpsCircle);

	// Some decoration
	serie->SetBrush(*wxYELLOW);
	serie->SetSymbol(mpsCircle);
~~~~

If I change the above line to `serie2->AddData(i / 10.0 + 5, 3 * sin(i / 10.0), true);`  (note there is a `3`), here is the result:

![image](https://github.com/GitHubLionel/wxMathPlot/assets/561818/3a461bd0-2364-44bb-907f-1c0d419334aa)

Looks nice.

About the tick lines in the second image, I think it is expected. Maybe, we can find a way to make those tick lines overlap (not interleaved)

**Comment by GitHubLionel** [comment]
Hi,
Yes looks good for this simple case.
For the ticks lines, there is no reason for the lines overlap since scales are different over the two axis. Solution is to not draw lines on the second axis. I add "show grid" option on configuration menu on new commit.

**Comment by asmwarrior** [comment]
> Hi, Yes looks good for this simple case. For the ticks lines, there is no reason for the lines overlap since scales are different over the two axis. Solution is to not draw lines on the second axis. I add "show grid" option on configuration menu on new commit.

OK, I understand.
I see your improvement. It works fine.

**Comment by GitHubLionel** [comment]
Yes seems to work fine.
A collateral effect is that I can add axis in the configuration menu.
See new commit.

**Comment by asmwarrior** [comment]
> Yes seems to work fine. A collateral effect is that I can add axis in the configuration menu. See new commit.

Thanks, I just test your recent commit.

The steps are:

1, add a X and Y axis
2, add two curves(curve0 and curve1), they both on Y axis
3, show the plot
4, in the config dialog, add a new Y axis, and set it as a Y2 axis in the right
5, set the curve1 to using Y2 axis
6, fit the plot
7, the result is good!
8, I also see when using the "track" feature, and when mouse is hover near the curve1, the correct coordinates of curve1 is shown.

Thanks!

I think this issue is implemented. Great work.

**Comment by GitHubLionel** [comment]
Thanks.
Just another small correction when we rename an axis to update the choice list.
If all is ok, I will merge with the master branch.

**Comment by asmwarrior** [comment]
> Thanks. Just another small correction when we rename an axis to update the choice list. If all is ok, I will merge with the master branch.

When testing, I see I can rename the Y2 axis name without any issue.

**Comment by GitHubLionel** [comment]
I think that my boolean FitWithBound is no more needed.
If you comment lines 2352-2357, it still work
```
//  if (FitWithBound)
//  {
//    m_posY2 += (m_posY - old_posY) / Y2Factor;
//    FitWithBound = false;
//  }
//  else```

**Comment by asmwarrior** [comment]
Like the below code:
~~~~
  double old_posY = m_posY;
  m_posX = (rect.Xmin + rect.Xmax) / 2 - (m_plotWidth / 2 + m_margin.left) / m_scaleX;
  m_posY = (rect.Ymin + rect.Ymax) / 2 + (m_plotHeight / 2 + m_margin.top) / m_scaleY;
//  if (FitWithBound)
//  {
//    m_posY2 += (m_posY - old_posY)/Y2Factor;
//    FitWithBound = false;
//  }
//  else
    m_posY2 = (rect.Y2min + rect.Y2max) / 2 + (m_plotHeight / 2 + m_margin.top) / m_scaleY2;
~~~~

Yes, I just tested.
It still works because I see the `m_scaleY2` is always calculated by the `Ay/Ay2`.

I see that when "lock aspect" is selected, I see this Y2 scale feature is disabled. right?

**Comment by GitHubLionel** [comment]
> I see that when "lock aspect" is selected, I see this Y2 scale feature is disabled. right?
Yes, because "lock aspect" mean axis X and Y have the same scale, so Y2 logically has also the same scale. But we can decide that Y2 keep his own scale.

**Comment by asmwarrior** [comment]
> But we can decide that Y2 keep his own scale.

Logically, I think the current method is good, I mean they (Y, Y2 and X) should have same scales.

**Comment by GitHubLionel** [comment]
Yes that what I think.
I made more corrections, difficult to stop me ! lol
But I think that now the code is more clean.

**Comment by asmwarrior** [comment]
![image](https://github.com/GitHubLionel/wxMathPlot/assets/561818/3819c9cd-f7ea-4a8c-961f-7c0f1eca01da)


I'm testing the latest code, and see the image above. When the Y2 is enabled, and I "Fit" the canvas, I see that the X coordinates is not fit the whole horizontal space, the left side, half of the marker is truncated, in the right side, there is a space between the marker and the boarder. Any ideas?

Another issue is that the name "Serie", I think it is not a English name, maybe it is from French language. Any idea to change those names? 

Thanks.

**Comment by asmwarrior** [comment]
I think the English translation is the "series", so you have "series 0", "series 1" for different curves.

**Comment by GitHubLionel** [comment]
LIMIT increase a little the bound.
So I replace LIMIT by a property m_limit_percent. Then if you call SetLimitPercent(0); before added points, you will have end point exactly on the axis.

**Comment by asmwarrior** [comment]
In the function: `void MathPlotConfigDialog::Initialize()`

I have to use this else condition:

```
    else if (classname.IsSameAs(_T("mpScaleY")))
    {
      if (((mpScaleY*)axis)->IsY2Axis())
      {
        ChoiceAxis->Append(_T("Y2 axis - ") + axis->GetName());
      }
      else
        ChoiceAxis->Append(_T("Y axis - ") + axis->GetName());
    }
```

Because I have some `mpScale` derived class, which should not list in the axes list.

**Comment by asmwarrior** [comment]
I think I have found a bug of the fit function

~~~~
	// add a simple sinus serie
	mpFXYVector *serie = m_plot->GetXYSeries(0);
	for (int i = 0; i < 100; i++)
		serie->AddData(i / 10.0 + 40000, sin(i / 10.0), true);
	m_plot->Fit();  //  UpdateAll
	legend->SetNeedUpdate();

	mpFXYVector *serie2 = m_plot->GetXYSeries(1);
    for (int i = 0; i < 100; i++)
		serie2->AddData(i / 10.0 + 40000 + 5, 3 * sin(i / 10.0), true);
    //serie2->SetY2Axis(true);
    serie2->SetBrush(*wxRED);
	serie2->SetSymbol(mpsCircle);

	// Some decoration
	serie->SetBrush(*wxYELLOW);
	serie->SetSymbol(mpsCircle);
~~~~

When started, I see the plot looks like below:

![image](https://github.com/GitHubLionel/wxMathPlot/assets/561818/9cf80417-d7e8-4273-b24d-7f85baa99f2f)


When I use the mouse to zoom to a rectangle, I got the correct result:

![image](https://github.com/GitHubLionel/wxMathPlot/assets/561818/b77a1b50-14ec-4bfa-b6ab-c4ceed4eb9e5)

This is the latest git head code of the demo project.

**Comment by GitHubLionel** [comment]
This is not a bug. Result is correct, this the effect of LIMIT now renamed m_limit_percent.
Add serie->SetLimitPercent(0); and serie2->SetLimitPercent(0); and you will have a more precise fit.
```  mpFXYVector *serie = m_plot->GetXYSeries(0);
  serie->SetLimitPercent(0);
  for (int i = 0; i < 100; i++)
    serie->AddData(i / 10.0 + 40000, sin(i / 10.0), true);
  m_plot->Fit();  //  UpdateAll
  legend->SetNeedUpdate();

  mpFXYVector *serie2 = m_plot->GetXYSeries(1);
  serie2->SetLimitPercent(0);
    for (int i = 0; i < 100; i++)
    serie2->AddData(i / 10.0 + 40000 + 5, 3 * sin(i / 10.0), true);
    //serie2->SetY2Axis(true);
    serie2->SetBrush(*wxRED);
  serie2->SetSymbol(mpsCircle);`

**Comment by asmwarrior** [comment]
OK, thanks for the help. I remembered that I have reported such issue before, see here:

[bugs in calculation the bounding box · Issue #18 · GitHubLionel/wxMathPlot](https://github.com/GitHubLionel/wxMathPlot/issues/18)

I think we should consider the actual x or y's range, not the values. 

For example, if the x coordinates range is [40000, 40020], it should be the same as the [0, 20], because they have the same range. But what I reported in that issue is that currently, the gap is depends on the values, not one the ranges.

**Comment by asmwarrior** [comment]
I read the code how to calculate the boundary of the x or y.

I see the logic is that, when you add each point, you need to calculate the bounding box. When calculate the bounding box's x limit(the minimal x or maximal x), you have use the m_limit_percent.

I think this logic is not correct. When calculating the bounding box, we should NOT use the m_limit_percent, as you mentioned, you should make m_limit_percent == 0.

But when you run the fit function, you have to expand the bounding box a little. So, for a line which has range [40000, 40020], you may has an actual range `[40000 - 20*5%, 400020 + 20*5%]`, here the `20` is the range length, and `5%` is the actual `m_limit_percent` we should set.

I hope you can understand my idea. Thanks.

**Comment by GitHubLionel** [comment]
Ok, I try to explain why I use this method. 
In my application, the points are not calculated in one time like in your sinus sample. They are computed step by step and added to the serie. To avoid total repainting each time I add new point, I try to anticipate the next point by increasing the bound of the added point.

**Comment by asmwarrior** [comment]
> Ok, I try to explain why I use this method. In my application, the points are not calculated in one time like in your sinus sample. They are computed step by step and added to the serie. To avoid total repainting each time I add new point, I try to anticipate the next point by increasing the bound of the added point. An idea I just have maybe better is to compute the delta between previous point added and add this delta to the bond ...


If I understand your idea, this is some kind of Kalman filter, but you still need a delta,  not the absolute valve of the point.

I think my idea does not conflict with your idea,  The delta is just a range.  So you need to keep some space in the margin,  In this case when new point added you don't need to repaint the whole plot chat,  You just need to draw the new point.

**Comment by GitHubLionel** [comment]
Ok, I try a new approach with delta on X and Y.
It seem to work.

**Comment by asmwarrior** [comment]
> Ok, I try a new approach with delta on X and Y. It seem to work.

I see your new approach, but I still don't know the logic.

If I have several points like:

~~~~
(1,2), (2,2), (3,2.5)
~~~~

In this case, the y delta could be "0"? Is that correct?

**Comment by asmwarrior** [comment]
I have such test code for the demo project:

~~~~
	mpScaleY *rightAxis = new mpScaleY(wxT("Y2"), mpALIGN_RIGHT, true, true);
	m_plot->AddLayer(rightAxis);
~~~~

and

~~~
	// add a simple sinus serie
	mpFXYVector *serie0 = m_plot->GetXYSeries(0);
    //serie0->SetLimitPercent(0);
	for (int i = 0; i < 100; i++)
		serie0->AddData(i / 10.0 + 40000, sin(i / 10.0), true);
    	// Some decoration
	serie0->SetBrush(*wxYELLOW);
	serie0->SetSymbol(mpsCircle);

	mpFXYVector *serie1 = m_plot->GetXYSeries(1);
    //serie1->SetLimitPercent(0);
    for (int i = 0; i < 100; i++)
		serie1->AddData(i / 10.0 + 40000 + 5, 3 * sin(i / 10.0) + 4.0, true);
    serie1->SetBrush(*wxRED);
	serie1->SetSymbol(mpsCircle);
	//serie1->SetY2Axis(true);

    mpFXYVector *serie2 = m_plot->GetXYSeries(2);
    //serie2->SetLimitPercent(0);
    for (int i = 0; i < 100; i++)
		serie2->AddData(i / 10.0 + 40000 + 5, -4.0, true);
    serie2->SetBrush(*wxBLUE);
	serie2->SetSymbol(mpsSquare);
    //serie2->SetY2Axis(true);

	legend->SetNeedUpdate();
	m_plot->Fit();  //  UpdateAll
~~~

You see, by default, all the 3 lines were added to the Y axis, and there is NO lines in the Y2 axis.

When start up, it looks like below:

![image](https://github.com/GitHubLionel/wxMathPlot/assets/561818/d2464756-d326-4d55-beba-38ba851465d1)

This looks OK.

Now, in the configure dialog, I try to select the "serial 1" to the Y2 axis, see the below image:

![image](https://github.com/GitHubLionel/wxMathPlot/assets/561818/ed2f099d-e76c-4e27-b0d7-f2d2163fa5db)

After "Fit" the screen, the result looks wrong here, see the image below:

![image](https://github.com/GitHubLionel/wxMathPlot/assets/561818/8f60b199-81e8-4ec1-8748-11fe9e42788a)

You can see the arrow in the above image, it should span the whole Y2 ranges.

**Comment by GitHubLionel** [comment]
Ok, I add a test on UpdateBBox.
Note that I see that I must correct the "Take care of scale : restrict bound" part.

**Comment by asmwarrior** [comment]
> Ok, I add a test on UpdateBBox. 

![image](https://github.com/GitHubLionel/wxMathPlot/assets/561818/dd869346-deab-415a-b015-5c05ebe4001e)

I update to your current git head, and it looks OK now. 

Thanks.

> Note that I see that I must correct the "Take care of scale : restrict bound" part.

You mean you need to update/correct some code in the bounding box calculation?

**Comment by asmwarrior** [comment]
![boundingbox](https://github.com/GitHubLionel/wxMathPlot/assets/561818/6dee8177-d7f2-40f0-aa3d-0f597c51e128)

I draw an image of how the bounding box can be calculated.

We should keep the actual bounding box(the red dashed rectangle), and the outer blue rectangle is the window/chart size. There is a gap, such as 10% between those two rectangles.

So, the 

`m_extended_minX = m_actual_minX - (m_actual_range_X * m_limit_percent);`

Here, we could still use the `m_limit_percent`, in my case, it is 10%.

I hope you can understand my idea.

When adding one data point, the `m_actual_minX` should be updated, and extended by `m_limit_percent`.

**Comment by asmwarrior** [comment]
How to draw a customized class object in the Y2 axis related coordinates.

Hi, I have a customized class like below:

~~~
class WXDLLIMPEXP_MATHPLOT HorizontalLine : public mpScale
{
public:
    HorizontalLine(const wxString& name = _T("label"), int flags = mpALIGN_CENTERY, bool grids = false)
        : mpScale(name, flags, grids)
    {
    }

    /** Layer plot handler.
     * This implementation will plot the ruler adjusted to the visible area.
     */
    virtual void DoPlot(wxDC& dc, mpWindow& w);

    /** Specifies that this is a Scale layer.
     * @return always \a TRUE
     * @sa mpLayer::IsScale
     */
    virtual bool IsScale(mpScaleType* scale)
    {
        *scale = NULL;
        return true;
    }

    virtual bool IsLogAxis();
    virtual void SetLogAxis(bool log);
    virtual int GetOrigin(mpWindow &w);
    virtual void DrawScaleName(wxDC &dc, mpWindow &w, int origin, int labelSize);

    /// the y coordinate of the horizontal line
    int m_Y = 0;

protected:

    DECLARE_DYNAMIC_CLASS(MyScaleX)
};
~~~

This object draws like X axis, and has a m_Y field, so I can draw a horizontal line.

But what I want is that I would like to draw this horizontal line in the Y2's coordinates system.

I see that if I want to draw on a Y2's coordinates, I have to derive this class from the `mpFunction`, and later, those objects will exists in the legend's curve list. Since this is a custom class, I don't want to show this object in the legend. How to do that? Thanks.

**Comment by GitHubLionel** [comment]
> You mean you need to update/correct some code in the bounding box calculation?

Yes because we have the choice between fixed or automatic bound. I need to update the code for Y2.

If we want extra marge, the more simple is to add this marge when we draw the box around the curve. For example, if we want 5 pixels, then in mpWindow::OnPaint method :
```
  // Draw background plot area
  trgDc->SetBrush(m_bgColour);
  trgDc->SetTextForeground(m_fgColour);
  trgDc->DrawRectangle(m_margin.left-5, m_margin.top-5, m_plotWidth+10, m_plotHeight+10);
```

For an horizontal line, why not use mpFX :
```
class HorizontalLine : public mpFX
{
    double m_yvalue;
  public:
    HorizontalLine (double yvalue) :
        mpFX(wxT("Horizontal line"), mpALIGN_LEFT)
    {
      m_yvalue = yvalue;
      wxPen FXpen(*wxGREEN, 1, wxPENSTYLE_SOLID);
      SetDrawOutsideMargins(false);
      SetContinuity(true);
      SetPen(FXpen);
      SetY2Axis(true);
    }
    virtual double GetY(double WXUNUSED(x))
    {
      return m_yvalue;
    }
    virtual double GetMinY()
    {
      return m_yvalue - 0.05;
    }
    virtual double GetMaxY()
    {
      return m_yvalue + 0.05;
    }
};
```
Then to add layer horizontal line :
`m_plot->AddLayer(new HorizontalLine(3.0));`

**Comment by asmwarrior** [comment]
Hi, about the HorizontalLine class design, thanks.

It's working now, here is the code that I derived from mpFX

~~~
IMPLEMENT_DYNAMIC_CLASS(HorizontalLine, mpFunction)

void HorizontalLine::DoPlot(wxDC& dc, mpWindow& w)
{
    int orgy = GetOrigin(w);
    wxCoord iy = w.y2p(m_Y, UseY2Axis);

    // Draw nothing if we are outside margins
    if (!m_drawOutsideMargins && ((iy > (w.GetScreenY() - w.GetMarginBottom())) || (iy < w.GetMarginTop())))
      return;

    // Draw horizontal lines from boundary minX to boundary maxX
    dc.DrawLine(m_plotBondaries.startPx, iy, m_plotBondaries.endPx, iy);
}
~~~

Those lines are like threshold lines.

Also, I have disable the bounding box calculation:

```
    virtual bool HasBBox() override
    {
      return false;
    }
```

And the type I set is: `m_type = mpLAYER_INFO;`. So they are not shown in the legend.

**Comment by GitHubLionel** [comment]
Ok, I see what you want. Simply is to use mpFunction.
I add mpHorizontalLine and mpVerticalLine.
Now the only code you have to do is : `m_plot->AddLayer(new mpHorizontalLine(3.0, *wxGREEN, true));`

**Comment by asmwarrior** [comment]
> Ok, I see what you want. Simply is to use mpFunction. I add mpHorizontalLine and mpVerticalLine. Now the only code you have to do is : `m_plot->AddLayer(new mpHorizontalLine(3.0, *wxGREEN, true));`

Thanks, the new class works OK, See the image below, it is shown in the Y2 coordinates.

![image](https://github.com/GitHubLionel/wxMathPlot/assets/561818/b0e71fa2-b8c8-4433-86a9-a5918b519a2f)

And here is the related code:

~~~
	// add a simple sinus serie
	mpFXYVector *serie0 = m_plot->GetXYSeries(0);
    //serie0->SetLimitPercent(0);
	for (int i = 0; i < 100; i++)
		serie0->AddData(i / 10.0 + 40000, sin(i / 10.0), true);
    	// Some decoration
	serie0->SetBrush(*wxYELLOW);
	serie0->SetSymbol(mpsCircle);

	mpFXYVector *serie1 = m_plot->GetXYSeries(1);
    for (int i = 0; i < 100; i++)
		serie1->AddData(i / 10.0 + 40000 + 5, 3 * sin(i / 10.0) + 4.0, true);
    serie1->SetBrush(*wxRED);
	serie1->SetSymbol(mpsCircle);
	serie1->SetY2Axis(true);

    mpFXYVector *serie2 = m_plot->GetXYSeries(2);
    for (int i = 0; i < 100; i++)
		serie2->AddData(i / 10.0 + 40000 + 5, -4.0, true);
    serie2->SetBrush(*wxBLUE);
	serie2->SetSymbol(mpsSquare);
    //serie2->SetY2Axis(true);


    mpHorizontalLine* hLine = new mpHorizontalLine(/*y*/ 6, /*color*/ *wxGREEN, /*useY2Axis*/ true);
    wxPen FXpen(*wxGREEN, 3, wxPENSTYLE_SOLID);
    hLine->SetPen(FXpen);
    m_plot->AddLayer(hLine);

	legend->SetNeedUpdate();
	m_plot->Fit();  //  UpdateAll
~~~

One minor question: is it possible to draw this horizontal line in the background as the "axis" does. I mean when drawing all the layer objects, those horizontal lines should be drawn before the curves.

**Comment by GitHubLionel** [comment]
> One minor question: is it possible to draw this horizontal line in the background as the "axis" does. I mean when drawing all the layer objects, those horizontal lines should be drawn before the curves.

Simply add horizontal line layer before series layer.

**Comment by GitHubLionel** [comment]
Ok, I add z-order for plot layer. Not sure of order but we must not break the original order plot.
Also, replace color by pen in horizontal and vertical axis, so code is simply :
```
wxPen Hor_pen(*wxGREEN, 3, wxPENSTYLE_SOLID);
  m_plot->AddLayer(new mpHorizontalLine(5.0, Hor_pen, true));`

**Comment by asmwarrior** [comment]
> Ok, I add z-order for plot layer. Not sure of order but we must not break the original order plot. Also, replace color by pen in horizontal and vertical axis, so code is simply :
> 
> ```
> wxPen Hor_pen(*wxGREEN, 3, wxPENSTYLE_SOLID);
>   m_plot->AddLayer(new mpHorizontalLine(5.0, Hor_pen, true));`
> ```

Thanks, it works OK now. 

Another minor issue:

![image](https://github.com/GitHubLionel/wxMathPlot/assets/561818/189e10ab-ccfd-49c0-9942-577688ff3137)

The circle and square are using the same size = 6, but they look different.

~~~
    case mpsCircle:
      dc.DrawCircle(x, y, m_symbolSize);
      break;

    case mpsSquare:
      dc.DrawRectangle(x - m_symbolSize2, y - m_symbolSize2, m_symbolSize, m_symbolSize);
      break;
~~~

It looks like the circle has radius = 6, but for square, the side = 6.

**Comment by GitHubLionel** [comment]
> It looks like the circle has radius = 6, but for square, the side = 6.

Yes it is correct.

**Comment by asmwarrior** [comment]
OK, I think I'm going to close this issue, because all the feature I requested is completed.

Thanks.

If I have other questions or issues, I think will create new issues.
